distributed/worker.py

```diff
@@ -3445,47 +3461,78 @@ async def find_missing(self) -> None:
                 self.scheduler.who_has,
                 keys=[ts.key for ts in self._missing_dep_flight],
             )
-            self.update_who_has(who_has)
-            recommendations: Recs = {}
+            recommendations, instructions = self._update_who_has(
+                who_has, stimulus_id=stimulus_id
+            )
             for ts in self._missing_dep_flight:
                 if ts.who_has:
+                    assert ts not in recommendations
                     recommendations[ts] = "fetch"
             self.transitions(recommendations, stimulus_id=stimulus_id)
+            self._handle_instructions(instructions)
 
         finally:
             # This is quite arbitrary but the heartbeat has scaling implemented
             self.periodic_callbacks[
                 "find-missing"
             ].callback_time = self.periodic_callbacks["heartbeat"].callback_time
 
-    def update_who_has(self, who_has: dict[str, Collection[str]]) -> None:
-        try:
-            for dep, workers in who_has.items():
-                if not workers:
-                    continue
+    def _update_who_has(
+        self, who_has: dict[str, Collection[str]], *, stimulus_id: str
+    ) -> RecsInstrs:
+        recs: Recs = {}
+        instructions: Instructions = []
 
-                if dep in self.tasks:
-                    dep_ts = self.tasks[dep]
-                    if self.address in workers and self.tasks[dep].state != "memory":
-                        logger.debug(
-                            "Scheduler claims worker %s holds data for task %s which is not true.",
-                            self.name,
-                            dep,
+        for key, workers in who_has.items():
+            ts = self.tasks.get(key)
+            if not ts:
+                continue
+            workers = set(workers)
+
+            if self.address in workers:
+                workers.remove(self.address)
+                if ts.state != "memory":
+                    logger.debug(
+                        "Scheduler claims worker %s holds data for task %s, "
+                        "which is not true.",
+                        self.name,
+                        ts,
+                    )
+                    instructions.append(
+                        MissingDataMsg(
+                            key=key, errant_worker=self.address, stimulus_id=stimulus_id
                         )
```

Comment:
Do we have a reproducer for this case? I remember debugging a state like this recently and it turned out to be a broken transition on the scheduler side or some race condition.

Reply:
Yes, there is one in this PR: test_self_denounce_missing_data
If you remove the message (which reverts back to the behaviour of main), you get a deadlock.